2000-08-11  Nathan Sidwell  <nathan@codesourcery.com>

	* decl.c (check_tag_decl): Diagnose typename's which don't
	declare anything.

2001-07-25  Jakub Jelinek  <jakub@redhat.com>

	* g++.old-deja/g++.other/typename2.C: New test.

--- gcc/cp/decl.c.jj	Tue Jul 24 20:41:10 2001
+++ gcc/cp/decl.c	Wed Jul 25 17:02:34 2001
@@ -7067,7 +7067,8 @@ check_tag_decl (declspecs)
 	{
 	  ++found_type;
 
-	  if (IS_AGGR_TYPE (value) || TREE_CODE (value) == ENUMERAL_TYPE)
+	  if ((TREE_CODE (value) != TYPENAME_TYPE && IS_AGGR_TYPE (value))
+	      || TREE_CODE (value) == ENUMERAL_TYPE)
 	    {
 	      my_friendly_assert (TYPE_MAIN_DECL (value) != NULL_TREE, 261);
 	      t = value;
--- gcc/testsuite/g++.old-deja/g++.other/typename2.C.jj	Fri Feb 16 21:16:31 2001
+++ gcc/testsuite/g++.old-deja/g++.other/typename2.C	Wed Jul 25 17:11:23 2001
@@ -0,0 +1,7 @@
+// Build don't link:
+
+template<typename T>
+void f()
+{
+  typename T::u;  // ERROR - declare anything
+}
